File: backend/src/main/java/com/gxs/bank/controller/NotificationController.java
package com.gxs.bank.controller;

import com.gxs.bank.dto.response.ApiResponse;
import com.gxs.bank.model.Notification;
import com.gxs.bank.model.User;
import com.gxs.bank.service.NotificationService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/notifications")
@RequiredArgsConstructor
public class NotificationController {
    private final NotificationService notificationService;

    @GetMapping
    public ResponseEntity<ApiResponse<List<Notification>>> getNotifications(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(ApiResponse.ok(notificationService.getNotifications(user.getId())));
    }

    @GetMapping("/unread-count")
    public ResponseEntity<ApiResponse<Long>> getUnreadCount(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(ApiResponse.ok(notificationService.getUnreadCount(user.getId())));
    }

    @PutMapping("/{id}/read")
    public ResponseEntity<ApiResponse<String>> markAsRead(@AuthenticationPrincipal User user, @PathVariable UUID id) {
        notificationService.markAsRead(id);
        return ResponseEntity.ok(ApiResponse.ok("Marked as read"));
    }
}

File: backend-python/com/gxs/bank/controller/NotificationController.py
from fastapi import APIRouter, Depends

from com.gxs.bank.config.database import get_db
from com.gxs.bank.config.serialization import serialize
from com.gxs.bank.dto.response.ApiResponse import ok
from com.gxs.bank.security.JwtAuthenticationFilter import get_current_user
from com.gxs.bank.service.NotificationService import NotificationService


router = APIRouter(prefix="/api/notifications", tags=["Notification"])


@router.get("")
def getNotifications(current_user=Depends(get_current_user), db=Depends(get_db)):
    notifications = NotificationService(db).getNotifications(current_user.id)
    return ok(serialize(notifications))


@router.get("/unread-count")
def getUnreadCount(current_user=Depends(get_current_user), db=Depends(get_db)):
    count = NotificationService(db).getUnreadCount(current_user.id)
    return ok(count)


@router.put("/{notification_id}/read")
def markAsRead(notification_id: str, current_user=Depends(get_current_user), db=Depends(get_db)):
    NotificationService(db).markAsRead(notification_id)
    return ok("Marked as read")
